include/albatross/src/models/nearest_neighbor.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2019 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_SRC_MODELS_NEAREST_NEIGHBOR_MODEL_HPP_
+#define ALBATROSS_SRC_MODELS_NEAREST_NEIGHBOR_MODEL_HPP_
+
+namespace albatross {
+
+template <typename DistanceMetric> class NearestNeighborModel;
+
+template <typename FeatureType> struct NearestNeighborFit;
+
+template <typename FeatureType> struct Fit<NearestNeighborFit<FeatureType>> {
+
+  Fit() : training_data(){};
+
+  Fit(const RegressionDataset<FeatureType> &dataset) : training_data(dataset){};
+
+  bool operator==(const Fit<NearestNeighborFit<FeatureType>> &other) const {
+    return training_data == other.training_data;
+  }
+
+  RegressionDataset<FeatureType> training_data;
+};
+
+template <typename DistanceMetric>
+class NearestNeighborModel
+    : public ModelBase<NearestNeighborModel<DistanceMetric>> {
+
+public:
+  NearestNeighborModel() : distance_metric(){};
+
+  std::string get_name() const { return "nearest_neighbor_model"; };
+
+  template <typename FeatureType>
+  Fit<NearestNeighborFit<FeatureType>>
+  _fit_impl(const std::vector<FeatureType> &features,
+            const MarginalDistribution &targets) const {
+    return Fit<NearestNeighborFit<FeatureType>>(
+        RegressionDataset<FeatureType>(features, targets));
+  }
+
+  template <typename FeatureType>
+  auto fit_from_prediction(const std::vector<FeatureType> &features,
+                           const JointDistribution &prediction) const {
+    const NearestNeighborModel<DistanceMetric> m(*this);
+    MarginalDistribution marginal_pred(
+        prediction.mean, prediction.covariance.diagonal().asDiagonal());
```

Comment:
Wouldn't `prediction.covariance` work here? Or are you looking to zero the non-diagonal elements?

Reply:
Yeah exactly, I need to zero the non-diagonal elements since the `NearestNeighbor` model can never actually predict off diagonals.